### PR TITLE
fix: unsafe methods not causing cache purge

### PR DESCRIPTION
**Pull request by flakey5**
Fixes bug introduced in #3733 where unsafe methods no longer make it to the cache handler and cause a cache purge for an origin.

Also removes a duplicate test file.

cc @mcollina @Uzlopak

### CODE REVIEW

**Review by mcollina — APPROVED** [review]
lgtm

**Review by Uzlopak — CHANGES_REQUESTED** [review]
have to review

File: lib/interceptor/cache.js [hunk 1]
```diff
@@ -32,7 +32,7 @@ module.exports = (opts = {}) => {
 
   return dispatch => {
     return (opts, handler) => {
-      if (!opts.origin || !methods.includes(opts.method)) {
+      if (!opts.origin || (opts.method in util.safeHTTPMethods && !methods.includes(opts.method))) {
```
Comment (metcoder95) [inline lib/interceptor/cache.js:35]: Why not `includes`?

**Comment by Uzlopak** [comment]
I think there is a bigger issue here.

File: lib/interceptor/cache.js [hunk 2]
```diff
@@ -32,7 +32,7 @@ module.exports = (opts = {}) => {
 
   return dispatch => {
     return (opts, handler) => {
-      if (!opts.origin || !methods.includes(opts.method)) {
+      if (!opts.origin || (util.safeHTTPMethods.includes(opts.method) && !methods.includes(opts.method))) {
```
Comment (Uzlopak) [inline lib/interceptor/cache.js:35]: just delete this if block and your test will pass.

**Comment by Uzlopak** [comment]
According to my assessment the whole logic is flawed.

**Comment by Uzlopak** [comment]
provided an alternative PR #3739

**Comment by Uzlopak** [comment]
What is the point of "methods" if you ignore it basically?

**Comment by flakey5** [comment]
> What is the point of "methods" if you ignore it basically?

It isn't being ignored, [the check at the beginning of each request](https://github.com/flakey5/undici/blob/ab2d8ddbe979b6142e076ddbfded86e072b4fa21/lib/interceptor/cache.js#L35) is all it's needed for

**Comment by Uzlopak** [comment]
I thought about it and I think the methods option makes not much sense.

lets say you have array safeMethods with values [A,B] and methods with values [B]

So method is 

A => [A,B].includes(A) && !([B].includes(A)) => true && true => true
B => [A,B].includes(B) && !([B].includes(B)) => true && false => false
C => [A,B].includes(C) && !([B].includes(C)) => false && true => false

But what if I want to add in the future e.g. PROPFIND, which should be in theory a safe method (also ignore the fact that we would throw now a TypeError)

I add methods to PROPFIND but it wont work, because it is not in the safeMethods Array. I would purge the cache even if I added PROPFIND to the methods array.

**Comment by Uzlopak** [comment]
I think you partially convinced me.

File: test/cache-interceptor/interceptor.js [hunk 3]
```diff
@@ -1,240 +0,0 @@
-'use strict'
```
Comment (Uzlopak) [inline test/cache-interceptor/interceptor.js:1]: Why did you delete this test file?

**Comment by Uzlopak** [comment]
We could safe the two includes by generating ONE array

const safeMethodsToNotCache = util.safeHTTPMethods.filter(method => methods.includes(method) === false)

safeMethods would be [A,B], methods would be [B], so the resulting safeMethodToNotCache would be [A]

then the table would look like:

A => [A].includes(A) => true
B => [A].includes(B) => false
C => [A].includes(C) => false

File: test/interceptors/cache.js [hunk 4]
```diff
@@ -250,4 +250,55 @@ describe('Cache Interceptor', () => {
...
+    })
+
+    strictEqual(deleteByOriginCalled, false)
+
+    // Make sure unsafe methods cause a cache purge
+    await client.request({
+      origin: 'localhost',
+      method: 'DELETE',
+      path: '/'
+    })
```
Comment (Ethan-Arrowood) [inline test/interceptors/cache.js:300]: For coverage purposes, I'd like to see this repeated for all of the unsafe methods

**Review by Ethan-Arrowood — APPROVED** [review]
not sure why the `cache-interceptor/interceptor.js` was deleted. I'd prefer slightly more test coverage, but wont block on that. lgtm otherwise

**Comment by flakey5** [comment]
@Uzlopak @Ethan-Arrowood the deleted test file was a duplicate - it still exists under `test/interceptors/cache.js`

**Comment by Uzlopak** [comment]
please adapt the code regarding the one array includes check.

**Comment by flakey5** [comment]
> But what if I want to add in the future e.g. PROPFIND

I do think something like the `safeMethods` option you were talking about would be valuable for this, but I'm not sure how well that behavior falls into the scope of this pr. This pr (and the original caching pr) was mainly targeting the methods specified as safe in the http semantics spec ([9.2.1](https://www.rfc-editor.org/rfc/rfc9110#section-9.2.1-3))

I do think it'd be a good idea to make an issue or pr to talk about supporting methods that were defined in extension specs though

**Review by Uzlopak — APPROVED** [review]
LGTM

**Pull request merged**